feat(filter): allow permalinks after filtering for deeping-linking

goldsziggy opened the pull request:
Altered filter.tsx to allow altering the query-param on change.

In router.tsx I added the `onpopstate` function to handle back-buttons after filters where made.

for filter.tsx I opted for using the `window.history.replaceState` instead of the `window.history.pushState`, as pushState would have made the back-button go back to each letter as it was typed in the filter field.  Because I opted for the replaceState, I needed to add the `onpopstate` function to correctly handle back-button clicks again.

fix #153

herbrandson commented:
Thanks for the PR @goldsziggy . Sorry it's taken me a couple days to review it :(. This is looking good and I think it'll be a great feature!

I do have 2 asks on this if you don't mind...

First, we haven't been using hooks in k8dash to this point. I'm not entirely opposed to using them, but I'd like to avoid having a mix of the 2 paradigms. Would you mind converting this to be a component instead of using hooks?

Second, when we previously discussed updating the url on key-press, I didn't consider the effect that would have on the "back button". Looking at this PR, I'm wondering if it wouldn't be cleaner all around to only update the url on blur. In that case I don't think we'd have to worry about `onpopstate` and the back button would "just work™". Thoughts?

goldsziggy commented:
I think the onblur would be a great solution to this as well.  I think overall that would be more clear, as in my opinoin the `onpopstate` api isn't something that is seen very often and can quickly become a 'tech-debt-esque' kind of pattern.

I can take both of those updates on and update the PR

goldsziggy commented:
@herbrandson  So it looks like I still need to leverage the `onpopstate` to properly handle the back-button clicks.  I updated the PR to use lifecycle functions as opposed to hooks, and use the onblur as opposed to onchange for the filter button.

I also fixed up a quick behavior bug I saw in the sorter logic so I use the `field` prop now.

Clicking around seems fine to me.

Side note, glad to see the github lint check running!  We should do a PR bringing everything up-to-date, so it can be enforced.

herbrandson commented on client/src/router.tsx:
```diff
@@ -121,6 +121,8 @@ registerRoute('workload/statefulset/:namespace/:name', params => <StatefulSet {.
 
 window.addEventListener('hashchange', onNavigate);
 
+window.onpopstate = onNavigate; // this is used by filter/sort query params to handle back button clicks
```
For consistency, it would be nice to use `addEventListener` here instead (similar to the line above). Any objections?

herbrandson commented:
This is looking great @goldsziggy! I asked for one very small change, but after that I think we're good to merge.

Also, I just merged a change to fix the linter errors. Is you'd like to submit another PR to enforce it, that would be awesome.
